src/result.ts

```diff
@@ -308,3 +303,165 @@ export class Err<T, E> implements IResult<T, E> {
 }
 
 export const fromThrowable = Result.fromThrowable
+
+//#region Combine - Types
+
+// Given a union, this extracts the intersection of the unions by casting it
+// to a parameter of a function and then inferring from it.
+type IntersectOf<U extends unknown> = (U extends unknown ? (k: U) => void : never) extends (
+  k: infer I,
+) => void
+  ? I
+  : never
+
+// Given a union, this extracts the last possible value.
+type Last<U extends unknown> = IntersectOf<U extends unknown ? (x: U) => void : never> extends (
+  x: infer P,
+) => void
+  ? P
+  : never
+
+/// Given a union and a member, this excludes the member.
+type ExcludeUnionItem<U extends unknown, M extends unknown> = U extends M ? never : U
+
+// Given two types, this checks whether the A1 is extending the A2, including
+// the `never` type in calculation. If A1 is `never`, then this will be `0`.
+type IsExtending<A1 extends unknown, A2 extends unknown> = [A1] extends [never]
+  ? 0
+  : A1 extends A2
+  ? 1
+  : 0
+
+// Given an array and a type, this extends the array by appending the `A`.
+type Append<L extends ReadonlyArray<unknown>, A extends unknown> = [...L, A]
+
+// Given a union, this gives the array of the union members.
+type MemberListOf<Union, MemberList extends ReadonlyArray<unknown> = [], LastU = Last<Union>> = {
+  0: MemberListOf<ExcludeUnionItem<Union, LastU>, Append<MemberList, LastU>>
+  1: MemberList
+}[IsExtending<[Union], [never]>]
+
+// This is a helper type to prevent infinite recursion in typing rules.
+//
+// Use this with your `depth` variable in your types.
+type Prev = [
+  never,
+  0,
+  1,
+  2,
+  3,
+  4,
+  5,
+  6,
+  7,
+  8,
+  9,
+  10,
+  11,
+  12,
+  13,
+  14,
+  15,
+  16,
+  17,
+  18,
+  19,
+  20,
+  ...0[]
+]
```

Comment:
Yes, that's right. AFAIK, TypeScript already does recursion for only fixed amount of times and it is why we have 20 there.

Reply:
>  TypeScript already does recursion for only fixed amount of times

If TS already has a recursion limit, why would we then need to impose our own limit?

Reply:
TS by default tries to resolve the type recursively as much as possible and that usually ends up in hanging or simply replacing the type with `never` or `unknown` as you can expect. We just make this in that way so that the TS server will stop recursing indefinitely, i.e. not showing us "Type instantiation is excessively deep and possibly infinite" error.